autoPatchelfHook: add autoPatchelfIgnoreMissing

DavHau opened the pull request:
###### Motivation for this change
Currently autoPatchelfHook fails as soon as any dependency cannot be satisfied. But in some cases it is favorable to continue patching and ignore the missing dependencies. Some binary distributions contain optional features with their own dependencies. The current autoPatchelfHook forces you to patch all dependencies even if you know you will never use those features.
A specific example would be the python wheel distribution of Pytorch which contains files which depend on cuda, but in practice this dependency doesn't need to be satisfied if you don't plan to use cuda. (https://github.com/pytorch/pytorch/issues/38624)

To solve this problem, I added the possibility to ignore missing dependencies by setting the environment variable `autoPatchelfIgnoreMissing`.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

DavHau committed:
add autoPatchelfIgnoreMissingDeps to auto-patchelf.sh

FRidh commented:
Ignore missing what? I suggest appending something to the attribute that indicates what is missing.

aszlig commented on pkgs/build-support/setup-hooks/auto-patchelf.sh:
```diff
@@ -141,7 +141,7 @@ autoPatchelfFile() {
     # the stdenv setup script is run with set -e. The actual error is emitted
     # earlier in the previous loop.
-    [ $depNotFound -eq 0 ]
+    [ $depNotFound -eq 0 ] || [ ! -z $autoPatchelfIgnoreMissing ]
```
Just a small nitpick:

```suggestion
    [ $depNotFound -eq 0 -o -n "$autoPatchelfIgnoreMissing" ]
```

aszlig commented:
> Ignore missing what? I suggest appending something to the attribute that indicates what is missing.

@FRidh: I also found the naming a bit odd, but haven't found a better name that is to the point and also short. Something like `autoPatchelfIgnoreMissingSharedLibraryDependencies` clearly is a *bit* long and `apIgnoreMissingLibDeps` also sounds weird.

I think just "ignore missing" is okay in the context of autoPatchelf, since the only thing that could be missing are shared library dependencies. Maybe I'm missing something here, but what's the first thing that comes to your mind in this context and "something missing"?

FRidh commented:
Like you wrote, it seems odd. I was thinking `autoPatchelfIgnoreMissingDeps` or `autoPatchelfIgnoreMissingLibs`.

DavHau commented:
Added the proposed syntax change from @aszlig and renamed the variable to `autoPatchelfIgnoreMissingDeps`. I tested it again and it works like expected.
